File: declarative_sqlite/lib/src/builders/table_builder.dart
import 'package:declarative_sqlite/src/schema/db_column.dart';
import 'package:declarative_sqlite/src/builders/column_builder.dart';
import 'package:declarative_sqlite/src/builders/date_column_builder.dart';
import 'package:declarative_sqlite/src/builders/fileset_column_builder.dart';
import 'package:declarative_sqlite/src/builders/guid_column_builder.dart';
import 'package:declarative_sqlite/src/builders/integer_column_builder.dart';
import 'package:declarative_sqlite/src/builders/key_builder.dart';
import 'package:declarative_sqlite/src/builders/real_column_builder.dart';
import 'package:declarative_sqlite/src/builders/text_column_builder.dart';
import 'package:declarative_sqlite/src/schema/db_table.dart';

class TableBuilder {
  final String name;
  final List<ColumnBuilder> _columnBuilders = [];
  final List<KeyBuilder> _keyBuilders = [];

  TableBuilder(this.name);

  T _addColumn<T extends ColumnBuilder>(T builder) {
    _columnBuilders.add(builder);
    return builder;
  }

  GuidColumnBuilder guid(String name) => _addColumn(GuidColumnBuilder(name));
  TextColumnBuilder text(String name) => _addColumn(TextColumnBuilder(name));
  IntegerColumnBuilder integer(String name) =>
      _addColumn(IntegerColumnBuilder(name));
  RealColumnBuilder real(String name) => _addColumn(RealColumnBuilder(name));
  DateColumnBuilder date(String name) => _addColumn(DateColumnBuilder(name));
  FilesetColumnBuilder fileset(String name) =>
      _addColumn(FilesetColumnBuilder(name));

  KeyBuilder key(List<String> columns) {
    final builder = KeyBuilder(columns);
    _keyBuilders.add(builder);
    return builder;
  }

  DbTable build() {
    final columns = _columnBuilders.map((b) => b.build()).toList();
    final hlcColumns = <DbColumn>[];
    for (final column in columns) {
      if (column.isLww) {
        hlcColumns.add(
          DbColumn(
            name: '${column.name}__hlc',
            logicalType: 'hlc',
            type: 'TEXT',
            isNotNull: false, // Nullable to support partial inserts
            isParent: false,
            isLww: false,
          ),
        );
      }
    }

    final isSystemTable = name.startsWith('__');

    final systemColumns = isSystemTable
        ? <DbColumn>[]
        : [
            DbColumn(
              name: 'system_id',
              logicalType: 'guid',
              type: 'TEXT', // GUID
              isNotNull: true,
              defaultValue: '00000000-0000-0000-0000-000000000000',
              isParent: false,
              isLww: false,
            ),
            DbColumn(
              name: 'system_created_at',
              logicalType: 'hlc',
              type: 'TEXT', // HLC
              isNotNull: true,
              defaultValue: '000000000000000:000000000:000000000000000000000000000000000000', // HLC format: milliseconds(15):counter(9):nodeId(36)
              isParent: false,
              isLww: false,
            ),
            DbColumn(
              name: 'system_version',
              logicalType: 'hlc',
              type: 'TEXT', // HLC
              isNotNull: true,
              defaultValue: '000000000000000:000000000:000000000000000000000000000000000000', // HLC format: milliseconds(15):counter(9):nodeId(36)
              isParent: false,
              isLww: false,
            ),
          ];

    return DbTable(
      name: name,
      columns: [...systemColumns, ...columns, ...hlcColumns],
      keys: _keyBuilders.map((b) => b.build()).toList(),
    );
  }
}

File: packages/core/src/schema/builders/table-builder.ts
import type { DbTable, DbColumn } from '../types';
import {
  TextColumnBuilder,
  IntegerColumnBuilder,
  RealColumnBuilder,
  GuidColumnBuilder,
  DateColumnBuilder,
  FilesetColumnBuilder,
} from './column-builders';
import { KeyBuilder } from './key-builder';

type AnyColumnBuilder = 
  | TextColumnBuilder
  | IntegerColumnBuilder
  | RealColumnBuilder
  | GuidColumnBuilder
  | DateColumnBuilder
  | FilesetColumnBuilder;

/**
 * Builder for database tables
 */
export class TableBuilder {
  private columnBuilders: AnyColumnBuilder[] = [];
  private keyBuilders: KeyBuilder[] = [];
  
  constructor(public readonly name: string) {}
  
  /**
   * Add a text column
   */
  text(name: string): TextColumnBuilder {
    const builder = new TextColumnBuilder(name);
    this.columnBuilders.push(builder);
    return builder;
  }
  
  /**
   * Add an integer column
   */
  integer(name: string): IntegerColumnBuilder {
    const builder = new IntegerColumnBuilder(name);
    this.columnBuilders.push(builder);
    return builder;
  }
  
  /**
   * Add a real (float) column
   */
  real(name: string): RealColumnBuilder {
    const builder = new RealColumnBuilder(name);
    this.columnBuilders.push(builder);
    return builder;
  }
  
  /**
   * Add a GUID column
   */
  guid(name: string): GuidColumnBuilder {
    const builder = new GuidColumnBuilder(name);
    this.columnBuilders.push(builder);
    return builder;
  }
  
  /**
   * Add a date column
   */
  date(name: string): DateColumnBuilder {
    const builder = new DateColumnBuilder(name);
    this.columnBuilders.push(builder);
    return builder;
  }
  
  /**
   * Add a fileset column
   */
  fileset(name: string): FilesetColumnBuilder {
    const builder = new FilesetColumnBuilder(name);
    this.columnBuilders.push(builder);
    return builder;
  }
  
  /**
   * Define a key (primary, unique, or index)
   */
  key(...columnNames: string[]): KeyBuilder {
    const builder = new KeyBuilder(columnNames);
    this.keyBuilders.push(builder);
    return builder;
  }
  
  /**
   * Build the final table definition
   */
  build(): DbTable {
    const columns = this.columnBuilders.map(b => b.build());
    
    // Add HLC columns for LWW fields
    const hlcColumns: DbColumn[] = [];
    for (const col of columns) {
      if (col.lww) {
        hlcColumns.push({
          name: `${col.name}__hlc`,
          type: 'TEXT',
          notNull: false,
        });
      }
    }

    // Add system columns
    const systemColumns: DbColumn[] = [];
    if (!this.name.startsWith('__')) {
      systemColumns.push({
        name: 'system_id',
        type: 'GUID',
        notNull: true,
        defaultValue: '00000000-0000-0000-0000-000000000000',
      });
      
      const zeroHlc = '000000000000000:000000000:000000000000000000000000000000000000';
      
      systemColumns.push({
        name: 'system_created_at',
        type: 'TEXT',
        notNull: true,
        defaultValue: zeroHlc,
      });
      
      systemColumns.push({
        name: 'system_version',
        type: 'TEXT',
        notNull: true,
        defaultValue: zeroHlc,
      });
      
      systemColumns.push({
        name: 'system_is_local_origin',
        type: 'INTEGER',
        notNull: true,
        defaultValue: 1,
      });
    }

    return {
      name: this.name,
      columns: [...systemColumns, ...columns, ...hlcColumns],
      keys: this.keyBuilders.map(b => b.build()),
      isSystem: this.name.startsWith('__'),
    };
  }
}
